bot/decorators.py

```diff
@@ -140,38 +105,35 @@ async def inner(self: Cog, ctx: Context, *args, **kwargs) -> None:
     return wrap
 
 
-def respect_role_hierarchy(target_arg: Union[int, str] = 0) -> Callable:
+def respect_role_hierarchy(member_arg: function.Argument) -> t.Callable:
     """
     Ensure the highest role of the invoking member is greater than that of the target member.
 
     If the condition fails, a warning is sent to the invoking context. A target which is not an
     instance of discord.Member will always pass.
 
-    A value of 0 (i.e. position 0) for `target_arg` corresponds to the argument which comes after
-    `ctx`. If the target argument is a kwarg, its name can instead be given.
+    `member_arg` is the keyword name or position index of the parameter of the decorated command
+    whose value is the target member.
 
     This decorator must go before (below) the `command` decorator.
     """
-    def wrap(func: Callable) -> Callable:
+    def decorator(func: t.Callable) -> t.Callable:
         @wraps(func)
-        async def inner(self: Cog, ctx: Context, *args, **kwargs) -> None:
-            try:
-                target = kwargs[target_arg]
-            except KeyError:
-                try:
-                    target = args[target_arg]
-                except IndexError:
-                    raise ValueError(f"Could not find target argument at position {target_arg}")
-                except TypeError:
-                    raise ValueError(f"Could not find target kwarg with key {target_arg!r}")
+        async def wrapper(*args, **kwargs) -> None:
+            log.trace(f"{func.__name__}: respect role hierarchy decorator called")
+
+            bound_args = function.get_bound_args(func, args, kwargs)
+            target = function.get_arg_value(member_arg, bound_args)
```

Comment:
I had to look around to see what `function` is. It gives the impression of being a parameter or a local variable, not an imported module. I'm not sure what a better name would be though.

Reply:
It could be qualified with `utils.` if you find that more readable.

Reply:
I have to import it in `utils/__init__.py` to support that. I tried it out and it felt too verbose, just cause of the length.

Reply:
`import function as func_utils` is another option, but feels redundant to do it everywhere the module is imported. On the other hand, it'd also feel redundant to actually name the module `func_utils` when it's already in the `utils` package.

Reply:
These are exactly the options I considered but didn't like any of them enough to propose them. Eh, it's fine the way it is, it's just the name `function` within the context of a decorator misleads, but it's not like you can't just look up where the name is defined.